src/rules.spec.ts

```diff
@@ -0,0 +1,323 @@
+      ],
+    };
+
+    const ruleWithNotOneOfCondition: Rule = {
+      conditions: [
+        {
+          operator: OperatorType.NOT_ONE_OF,
+          attribute: 'country',
+          value: ['USA', 'Canada'],
+        },
+      ],
+    };
+
+    const semverRule: Rule = {
+      conditions: [
+        {
+          operator: OperatorType.GTE,
+          attribute: 'version',
+          value: '1.0.0',
+        },
+        {
+          operator: OperatorType.LTE,
+          attribute: 'version',
+          value: '2.0.0',
+        },
+      ],
+    };
+    const ruleWithMatchesCondition: Rule = {
+      conditions: [
+        {
+          operator: OperatorType.MATCHES,
+          attribute: 'user_id',
+          value: '[0-9]+',
+        },
+      ],
+    };
+    const ruleWithNotMatchesCondition: Rule = {
+      conditions: [
+        {
+          operator: OperatorType.NOT_MATCHES,
+          attribute: 'user_id',
+          value: '[0-9]+',
+        },
+      ],
+    };
+    const subjectAttributes = {
+      totalSales: 50,
+      version: '1.5.0',
+      user_id: '12345',
+      country: 'USA',
+    };
+
+    it('should return true for a rule with empty conditions', () => {
+      expect(matchesRule(ruleWithEmptyConditions, subjectAttributes, false)).toBe(true);
+    });
+
+    it('should return true for a numeric rule that matches the subject attributes', () => {
+      expect(matchesRule(numericRule, subjectAttributes, false)).toBe(true);
+    });
+
+    it('should return false for a numeric rule that does not match the subject attributes', () => {
+      const failingAttributes = { totalSales: 101 };
+      expect(matchesRule(numericRule, failingAttributes, false)).toBe(false);
+    });
+
+    it('should return true for a rule with ONE_OF condition that matches the subject attributes', () => {
+      expect(matchesRule(ruleWithOneOfCondition, { country: 'USA' }, false)).toBe(true);
+    });
+
+    it('should return false for a rule with ONE_OF condition that does not match the subject attributes', () => {
+      expect(matchesRule(ruleWithOneOfCondition, { country: 'UK' }, false)).toBe(false);
+    });
+
+    it('should return false for a rule with ONE_OF condition when subject attribute is missing', () => {
+      expect(matchesRule(ruleWithOneOfCondition, { age: 10 }, false)).toBe(false);
+    });
+
+    it('should return false for a rule with NOT_ONE_OF condition that matches the subject attributes', () => {
+      expect(matchesRule(ruleWithNotOneOfCondition, { country: 'USA' }, false)).toBe(false);
+    });
+
+    it('should return true for a rule with NOT_ONE_OF condition that does not match the subject attributes', () => {
+      expect(matchesRule(ruleWithNotOneOfCondition, { country: 'UK' }, false)).toBe(true);
+    });
+
+    it('should return false for a rule with NOT_ONE_OF condition when subject attribute is missing', () => {
+      expect(matchesRule(ruleWithNotOneOfCondition, { age: 10 }, false)).toBe(false);
+    });
```

Comment:
Since this is tricky behavior, I'd like to also a see a test for `{ country: null }` (which is allowed by the signature `Record<string, any>`).

Reply:
Great callout